File: cpp/LeetCode2434.cpp
class Solution {
    public:
        string robotWithString(string s) {
            const string alphabet = "abcdefghijklmnopqrstuvwxyz";
            map<char, int> D = {};
            string t = "";
            string ans = "";
    
            // preprocessing
            for (char c : s) {
                D[c]++;
            }
            
            int i = 0;
            while (i < s.size()) {
                char c_min = find_min(alphabet, D);
    
                // check in t
                if (t != "" && t.at(t.size()-1) <= c_min) {
                    ans.push_back(t.at(t.size()-1));
                    t.pop_back();
                    continue;
                }
    
                // looking for c_min
                for (int j = i; j < s.size(); j++) {
                    char c = s.at(j);
                    if (c == c_min) {
                        ans.push_back(c);
                        D[c]--;
                        i++;
                        break;
                    } else {
                        t.push_back(c);
                        D[c]--;
                        i++;
                    }
                }
            }
    
            // attach t
            for (int k = t.size()-1; k >= 0; k--) {
                ans.push_back(t.at(k));
            }
    
            return ans;
        }
    
        char find_min(string alphabet, map<char, int> D) {
            for (char c : alphabet) {
                if (D[c] > 0) {
                    return c;
                }
            }
            return 'z';
        }
    
    };

File: python/LeetCode2434.py
class Solution:
    def robotWithString(self, s: str) -> str:
        alphabet = "abcdefghijklmnopqrstuvwxyz"
        D = dict()
        t = ""
        ans = ""
        for c in s:
            D[c] = D.get(c, 0) + 1

        i = 0
        while i < len(s):
            # Take the best I can do
            c_min = self.get_min(alphabet, D)

            # Check for t
            if t!="" and t[-1] <= c_min:
                ans += t[-1]
                t = t[:-1]
                continue

            # Find the best in s
            for j in range(i, len(s)):
                c = s[j]
                if c == c_min:
                    ans += c
                    D[c] -= 1
                    i += 1
                    break
                else:
                    t += c
                    D[c] -= 1
                    i += 1
        
        for i in range(len(t) - 1, -1, -1):
            ans += t[i]
        
        return ans

        
    def get_min(self, alphabet:str, D:dict):
        for c in alphabet:
            if D.get(c, 0) > 0:
                return c
        